source/extensions/filters/http/grpc_web/grpc_web_filter.cc

```diff
@@ -54,6 +74,82 @@ bool GrpcWebFilter::isGrpcWebRequest(const Http::RequestHeaderMap& headers) {
   return false;
 }
 
+bool GrpcWebFilter::isProtoEncodedGrpcWebResponseHeaders(
+    const Http::ResponseHeaderMap& headers) const {
+  // We expect the response headers to have 200 OK status (a valid gRPC, also gRPC-Web, response
+  // needs to have 200 OK status
+  // https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md#responses) and contain
+  // proto-encoded gRPC-Web content-type.
+  return Http::Utility::getResponseStatus(headers) == enumToInt(Http::Code::OK) &&
+         hasProtoEncodedGrpcWebContentType(headers);
+}
+
+bool GrpcWebFilter::hasProtoEncodedGrpcWebContentType(
+    const Http::RequestOrResponseHeaderMap& headers) const {
+  const Http::HeaderEntry* content_type = headers.ContentType();
+  if (content_type != nullptr) {
+    absl::string_view content_type_value = content_type->value().getStringView();
+    // We ignore "parameter" value. Note that "*( ";" parameter )" indicates that there can be
+    // multiple parameters.
+    absl::string_view current_content_type =
+        StringUtil::rtrim(content_type_value.substr(0, content_type_value.find_first_of(';')));
+    // We expect only proto encoding response
+    // https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-WEB.md. And the value of media-type is
+    // case-sensitive https://tools.ietf.org/html/rfc2616#section-3.7.
+    return StringUtil::CaseInsensitiveCompare()(
+               current_content_type, Http::Headers::get().ContentTypeValues.GrpcWebProto) ||
+           StringUtil::CaseInsensitiveCompare()(current_content_type,
+                                                Http::Headers::get().ContentTypeValues.GrpcWeb);
+  }
```

Comment:
Should this live in some shared utility? I know we infer other content types elsewhere, but I don't think we use this kind of logic. Should we?

Reply:
Can I add TODO here, I can have another PR focus on moving this code.

Reply:
I added a TODO here.